Add a simplistic C API for all unsigned data type widths as cell values to allow language interop

KitsuneAlex opened the pull request:
This would allow your library to be used in other languages which cannot directly interact with C++ through a simplistic C interface.

If you decide to merge this, it should be merged after [my last fix PR](https://github.com/facontidavide/Bonxai/pull/40).

Thanks for your excellent work on this project :)

KitsuneAlex commented:
Lightly tested on Windows x64, Linux x64/arm64 and macOS x64/arm64. Also confirmed working with tools like jnih and Kotlin CInterop :)

facontidavide commented:
Hi.

Thanks for taking this initiative! 

My suggestions are:

- please no change in the namespace ("boncai"). Keep it "bonxai"
- no alternative subdirectory. Instead, create a file called `bonxai_core/include/bonxai/bonxai.h` and add a target to the existing `bonxai_core/CMaleLists.txt`
- remove changesto "serialize"

KitsuneAlex commented:
This should do

KitsuneAlex commented:
I commented on the other closed PR btw if you didn't get notified, the print was your original code, i removed it in an earlier PR which was a mistake from my end, that's why i added it back. Your serilization will likely break without it, which i didn't realize before.

facontidavide commented on bonxai_core/include/bonxai/bonxai.h:
```diff
@@ -0,0 +1,133 @@
+#define BONXAI_OK 0
+#define BONXAI_ERR_INV_ARG 1
+
+#define BONXAI_FALSE 0
+#define BONXAI_TRUE 1
```
why?

KitsuneAlex replied:
As bool's and enum's are not exactly ideal for language interop scenarios because of unportable ABI an int type is usually employed, and if you have your boolean alias you might as well have fitting constants. These can be recognized by interop tools.

facontidavide commented on bonxai_core/src/pointer_input_stream.hpp:
```diff
@@ -0,0 +1,52 @@
+#include <streambuf>
+
+namespace Bonxai {
+struct PointerStreamBuffer final : public std::streambuf {
```
Add some doumentation. i don't know what the purpose of this is

KitsuneAlex replied:
Fixed

facontidavide commented on bonxai_core/src/vector_output_stream.hpp:
```diff
@@ -0,0 +1,75 @@
+#include <vector>
+
+namespace Bonxai {
+class VectorStreamBuffer final : public std::streambuf {
```
add documentation

KitsuneAlex replied:
Can do but this is an implementation detail, not sure if it needs _much_ documentation.

KitsuneAlex replied:
Fixed

KitsuneAlex commented on bonxai_core/include/bonxai/serialization.hpp:
```diff
@@ -81,6 +81,10 @@ inline void Serialize(std::ostream& out, const VoxelGrid<DataT>& grid) {
   char header[256];
   std::string type_name = details::demangle(typeid(DataT).name());
 
+  sprintf(
+      header, "Bonxai::VoxelGrid<%s,%d,%d>(%lf)\n", type_name.c_str(), grid.innetBits(),
+      grid.leafBits(), grid.voxelSize());
+
```
This was your code. I removed it in my previous PR as i assumed it was a debug line, but it wasn't. If you don't add this code back your serialization will be broken i think.

KitsuneAlex commented on bonxai_core/src/bonxai.cpp:
```diff
@@ -0,0 +1,198 @@
+BONXAI_DEFINE_GRID_IMPL(u8, uint8_t)
+BONXAI_DEFINE_GRID_IMPL(u16, uint16_t)
+BONXAI_DEFINE_GRID_IMPL(u32, uint32_t)
+BONXAI_DEFINE_GRID_IMPL(u64, uint64_t)
```
Done
